pyscraper/new_hansard.py

```diff
@@ -0,0 +1,1442 @@
+        'hs_8Clause',
+        'hs_newline10',
+        'hs_Question',
+    ]
+    indents = [
+        'hs_ParaIndent',
+        'hs_AmendmentLevel0'
+    ]
+    ignored_tags = [
+        'hs_6bPetitions',
+        'hs_76fChair',
+        'hs_3MainHdg',
+        'hs_Venue'
+    ]
+    root = None
+    ns = ''
+    ns_map = {}
+
+    debate_type = None
+    current_speech = None
+    date = ''
+    rev = 'a'
+    use_pids = True
+    current_col = 0
+    current_speech_num = -1
+    current_speech_part = 1
+    current_time = ''
+    output_heading = False
+    skip_tag = None
+    uc_titles = False
+
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        self.debate_type = None
+        self.current_speech = None
+        self.date = ''
+        self.rev = 'a'
+        self.current_col = 0
+        self.current_speech_num = 0
+        self.current_speech_part = 1
+        self.current_time = ''
+        self.root = None
+        self.input_root = None
+        self.output_heading = False
+
+    def get_tag_name_no_ns(self, tag):
+        # remove annoying namespace for brevities sake
+        tag_name = str(tag.tag)
+        tag_name = tag_name.replace(
+            '{{{0}}}'.format(self.ns),
+            ''
+        )
+        return tag_name
+
+    def get_pid(self):
+        if self.current_speech_num == -1:
+            self.current_speech_num = 0
+        return '{0}{1}.{2}/{3}'.format(
+            self.rev,
+            self.current_col,
+            self.current_speech_num,
+            self.current_speech_part
+        )
+
+    def get_speech_id(self):
+        if self.current_speech_num == -1:
+            self.current_speech_num = 0
+        return 'uk.org.publicwhip/{0}/{1}{2}.{3}.{4}'.format(
+            self.debate_type,
+            self.date,
+            self.rev,
+            self.current_col,
+            self.current_speech_num
+        )
+
+    def check_for_pi(self, tag):
+        pi = tag.xpath('.//processing-instruction("notus-xml")')
+        if len(pi) == 1:
+            self.parse_pi(pi[0])
+
+    # this just makes any gid redirection easier
+    def get_text_from_element(self, el):
+        text = self.get_single_line_text_from_element(el)
+        text = u'\n{0}\n'.format(text)
+        return text
+
+    def get_single_line_text_from_element(self, el):
+        text = u''.join(el.xpath('.//text()'))
+        text = re.sub('\n', ' ', text).strip()
+        return text
+
+    def clear_current_speech(self):
+        if self.current_speech is not None:
+            self.root.append(self.current_speech)
+        self.current_speech_part = 1
+        if self.current_speech_num == -1:
+            self.current_speech_num = 0
+        else:
+            self.current_speech_num = self.current_speech_num + 1
+        self.current_speech = None
+
+    def new_speech(self, member, url):
+        if self.current_speech is not None:
+            self.clear_current_speech()
```

Comment:
Can the checking whether current_speech is None or not be in this function? It's sometimes called with this check, sometimes out - is this purposeful?